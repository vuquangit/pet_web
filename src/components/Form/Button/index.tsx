import React, { ReactNode } from 'react'
import classNames from 'classnames'

interface ButtonProps {
  /**
   * Type button
   */
  type?: 'button' | 'submit' | 'reset' | undefined
  /**
   * Is this the principal call to action on the page?
   */
  primary?: boolean
  /**
   * What background color to use
   */
  backgroundColor?: string
  /**
   * How large should the button be?
   */
  size?: 'small' | 'medium' | 'large'
  /**
   * Button contents
   */
  label?: string
  /**
   * Optional click handler
   */
  onClick?: () => void

  /**
   * Class name
   */
  className?: string

  /**
   * Disable button
   */
  disabled?: boolean
  /**
   * Cypress cy
   */
  dataCy?: string
  children?: ReactNode
}

/**
 * Primary UI component for user interaction
 */
const Button = (props: ButtonProps) => {
  const {
    type = 'button',
    primary = false,
    size = 'medium',
    backgroundColor,
    label,
    className,
    dataCy,
    children,
    ...restProps
  } = props

  const buttonStyles = classNames(
    {
      'btn-default': !primary,
      'btn-primary': primary,
      'px-3 py-[2px] text-[12px]': size === 'small',
      'px-5 py-[4px] text-[14px]': size === 'medium',
      'px-7 py-[6px] text-[16px]': size === 'large',
    },
    className,
  )

  return (
    <button
      type={type}
      className={buttonStyles}
      style={{ backgroundColor }}
      data-cy={dataCy}
      {...restProps}
    >
      {label}
      {children}
    </button>
  )
}

export default Button
