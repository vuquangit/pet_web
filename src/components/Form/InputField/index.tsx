import React, { useState } from 'react'
import className from 'classnames'

import EyeIcon from '@/assets/icons/eye.svg'
import classNames from 'classnames'

interface PropType {
  /**
   * Label input
   */
  label?: string
  /**
   * Value input
   */
  value: string | number | readonly string[] | undefined
  /**
   * Placeholder
   */
  placeholder?: string
  /**
   * Type input
   */
  type?: React.HTMLInputTypeAttribute | undefined
  /**
   * Input name
   */
  name?: string
  /**
   * Autocomplete
   */
  autoComplete?: string
  /**
   * Input require
   */
  required?: boolean
  /**
   * Cypress cy
   */
  dataCy?: string
  /**
   * Disable input
   */
  disabled?: boolean
  /**
   * Read only
   */
  readOnly?: boolean
  /**
   * Event input change
   * @param value
   * @returns
   */
  onChange?: (value: string) => void
  onFocus?: React.FocusEventHandler<HTMLInputElement>
  classNameWrapper?: string
  classNameLabel?: string
  classNameInput?: string
}

const InputField: React.FC<PropType> = (props) => {
  const {
    name,
    label,
    type,
    dataCy,
    onChange,
    onFocus,
    classNameWrapper,
    classNameLabel,
    classNameInput,
    ...restProps
  } = props
  const [isShowPassword, setIsShowPassword] = useState<boolean>(false)

  const inputWrapperClass = className('relative', { 'mt-2': label })
  const eyeClass = className('h-4 w-4', {
    'fill-gray-400 ': !isShowPassword,
    'fill-gray-900': isShowPassword,
  })
  const inputClass = className(
    'input-default',
    {
      'input-error': restProps?.required,
      'bg-light-100 dark:bg-dark-100 text-light-200 dark:text-dark-200 border-light-100 dark:border-dark-100 cursor-not-allowed !shadow-[none]':
        props.readOnly,
    },
    classNameInput,
  )

  const onInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (!onChange) return
    onChange(e.target.value)
  }

  return (
    <div className={classNameWrapper}>
      <label
        htmlFor={name}
        className={classNames('block text-sm font-medium leading-6', classNameLabel)}
      >
        {label}
      </label>

      <div className={inputWrapperClass}>
        <input
          id={name}
          name={name}
          type={isShowPassword ? 'text' : type || 'text'}
          className={inputClass}
          onChange={onInputChange}
          onFocus={onFocus}
          data-cy={dataCy}
          {...restProps}
        />
        {type === 'password' && (
          <button
            type="button"
            className="absolute right-0 h-full px-3 -translate-y-1/2 top-1/2"
            onClick={() => setIsShowPassword(!isShowPassword)}
          >
            <EyeIcon className={eyeClass} />
          </button>
        )}
      </div>
    </div>
  )
}

export default InputField
