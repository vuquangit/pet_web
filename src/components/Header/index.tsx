import { useNavigate } from 'react-router-dom'

import NavigationApp from '@/components/Navigation'

import { storageKeys } from '@/constants/storage-keys'
import StorageService from '@/services/local-storage'
import { ROUTER_NAMES } from '@/constants/routerNames'
import { useAppDispatch, useAppSelector } from '@/store/hook'
import { resetCredentials } from '@/store/auth'
import LogoIcon from '@/assets/icons/logo.svg'
import ThemeSwitch from '../ThemeSwitch'

const Header: React.FC = () => {
  const navigate = useNavigate()
  const dispatch = useAppDispatch()
  const isAuthenticated = useAppSelector((state) => !!state.auth.role)

  const handleLogout = () => {
    console.log('Logout')
    StorageService.set(storageKeys.AUTH_PROFILE, null)
    dispatch(resetCredentials())
    navigate(ROUTER_NAMES.LOGIN)
  }

  return (
    <header className="flex items-center justify-between bg-white dark:bg-[#282829] px-5 py-3 dark:text-white shadow-md">
      <LogoIcon className="h-8" />

      {isAuthenticated && (
        <>
          <NavigationApp />
          <div className="flex">
            <ThemeSwitch />
            <button onClick={handleLogout}>Logout</button>
          </div>
        </>
      )}
    </header>
  )
}

export default Header
