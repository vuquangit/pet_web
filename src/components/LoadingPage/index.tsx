const LoadingPage = () => {
  return (
    <div>Loading page...</div>
  )
}

export default LoadingPage
