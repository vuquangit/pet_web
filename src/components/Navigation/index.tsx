import { ROUTER_NAMES } from '@/constants/routerNames'
import React from 'react'
import { Link } from 'react-router-dom'

interface Navigation {
  label: string
  path: string
}

const NavigationApp: React.FC = () => {
  const navList: Navigation[] = [
    {
      label: 'Home',
      path: ROUTER_NAMES.HOME,
    },
    {
      label: 'Conversations',
      path: ROUTER_NAMES.CONVERSATIONS,
    },
    {
      label: 'Friends',
      path: ROUTER_NAMES.FRIENDS,
    },
    {
      label: 'Groups',
      path: ROUTER_NAMES.GROUPS,
    },
  ]

  return (
    <nav>
      <ul className="flex gap-3">
        {navList.map((item) => (
          <li
            key={item.path}
            className="px-2"
          >
            <Link
              to={item.path}
              className="cursor-pointer text-black/80 hover:text-black dark:text-white/80 dark:hover:text-white"
            >
              {item.label}
            </Link>
          </li>
        ))}
      </ul>
    </nav>
  )
}

export default NavigationApp
