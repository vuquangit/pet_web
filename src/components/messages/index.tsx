import React, { FC, useState, useEffect } from 'react'
import { useParams } from 'react-router-dom'
import { get } from 'lodash'

import { RootState } from '@/store'
import { selectConversationById } from '@/store/conversations'
import { selectGroupById } from '@/store/group'
import { removeAllAttachments } from '@/store/message-panel'
import { addSystemMessage, clearAllMessages } from '@/store/system-messages'
import { useCreateMessageMutation } from '@/services/conversations'
import { getRecipientFromConversation } from '@/helpers'
import { useToast } from '@/hooks/useToast'
import { MessageAttachmentContainer } from './attachments/MessageAttachmentContainer'
import { MessageContainer } from './MessageContainer'
import { MessageInputField } from './MessageInputField'
import { MessagePanelHeader } from './MessagePanelHeader'
import { useAppSelector, useAppDispatch } from '@/store/hook'

type Props = {
  sendTypingStatus: () => void
  isRecipientTyping: boolean
}

export const MessagePanel: FC<Props> = ({ sendTypingStatus, isRecipientTyping }) => {
  const toastId = 'rateLimitToast'
  const dispatch = useAppDispatch()
  const { messageCounter } = useAppSelector((state: RootState) => state.systemMessages)
  const [content, setContent] = useState('')
  const { id: routeId = '' } = useParams()
  const user = useAppSelector((state: RootState) => state.auth)
  const { error } = useToast({ theme: 'dark' })
  const { attachments } = useAppSelector((state: RootState) => state.messagePanel)
  const conversation = useAppSelector((state: RootState) => selectConversationById(state, routeId))
  const group = useAppSelector((state: RootState) => selectGroupById(state, routeId))
  const selectedType = useAppSelector((state: RootState) => state.selectedConversationType.type)

  const recipient = getRecipientFromConversation(conversation, user)

  const [createMessage] = useCreateMessageMutation()

  useEffect(() => {
    return () => {
      dispatch(clearAllMessages())
      dispatch(removeAllAttachments())
    }
  }, [])

  const sendMessage = async () => {
    const trimmedContent = content.trim()
    if (!routeId) return
    if (!trimmedContent && !attachments.length) return
    const formData = new FormData()
    formData.append('id', routeId)
    trimmedContent && formData.append('content', trimmedContent)
    attachments.forEach((attachment: any) => formData.append('attachments', attachment.file))

    try {
      await createMessage({ id: routeId, type: selectedType, data: formData })
      // debugger
      setContent('')
      dispatch(removeAllAttachments())
      dispatch(clearAllMessages())
    } catch (err) {
      console.log('sendMessage error', err)

      // TODO: get status code
      const status = get(err, 'result.error.status')
      if (status === 429) {
        error('You are rate limited', { toastId })
        dispatch(
          addSystemMessage({
            id: messageCounter,
            level: 'error',
            content: 'You are being rate limited. Slow down.',
          }),
        )
      } else if (status === 404) {
        dispatch(
          addSystemMessage({
            id: messageCounter,
            level: 'error',
            content: 'The recipient is not in your friends list or they may have blocked you.',
          }),
        )
      }
    }
  }

  return (
    <div className="flex h-full w-full flex-col">
      <MessagePanelHeader />

      <div
        className="min-h-0 flex-1 overflow-auto"
        style={{ height: 'calc(100% - 600px)' }}
      >
        <MessageContainer
          isRecipientTyping={isRecipientTyping}
          recipient={recipient}
        />
      </div>

      <div className="m-4 rounded-[22px] border border-solid border-[#363636]">
        <MessageInputField
          content={content}
          setContent={setContent}
          sendMessage={sendMessage}
          sendTypingStatus={sendTypingStatus}
          placeholderName={
            selectedType === 'group' ? group?.title || 'Group' : recipient?.name || 'user'
          }
        />

        {attachments.length > 0 && <MessageAttachmentContainer />}
      </div>
    </div>
  )
}
