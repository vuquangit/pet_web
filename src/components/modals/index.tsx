import React, { FC, PropsWithChildren } from 'react'
// import {
//   ModalContainerStyle,
//   ModalContentBodyStyle,
//   ModalHeaderStyle,
// } from '../../utils/styles';

export const ModalHeader: FC<PropsWithChildren> = ({ children }) => {
  return (
    <header className="mt-5 flex w-full items-center justify-between px-[24px]">{children}</header>
  )
}

export const ModalContentBody: FC<PropsWithChildren> = ({ children }) => {
  return <div className="relative p-6">{children}</div>
}

type ModalContainerProps = {
  showModal?: boolean
}

export const ModalContainer: FC<PropsWithChildren & ModalContainerProps> = ({
  children,
  showModal,
}) => {
  console.log(`showModal: ${showModal}`)
  return (
    <div className="relative w-[650px] rounded-[10px] bg-white dark:bg-[#1f1f1f]">{children}</div>
  )
}
