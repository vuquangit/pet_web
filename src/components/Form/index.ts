import Button from './Button'
import InputField from './InputField'
import Skeleton from './Skeleton'

export { Button, InputField, Skeleton }
