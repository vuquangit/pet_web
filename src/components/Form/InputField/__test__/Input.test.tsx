import React from 'react'
import { render } from '@/tests/unit/utils'
import { describe, expect, it } from '@jest/globals'
import { screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { InputField } from '@/components/Form'

describe('Input Field component', () => {
  it('renders the component', async () => {
    const inputValue = ''
    const inputChange = jest.fn()

    const { container, getByDataCy, asFragment } = render(
      <InputField
        value={inputValue}
        label="Input label"
        dataCy="input-cy"
        onChange={inputChange}
        placeholder="Input placeholder"
      />,
    )

    const labelEl = container.querySelector('label')
    // @ts-ignore
    expect(labelEl).toHaveTextContent('Input label')

    const inputEL = container.querySelector('input')
    if (inputEL) {
      // fireEvent.change(inputEL, {target: {value: 'change value'}})

      await userEvent.type(screen.getByPlaceholderText(/Input placeholder/i), 'kunal')
      console.log('getByDataCy:', inputEL)
    }
    // @ts-ignore
    expect(getByDataCy('input-cy')).toHaveValue('')

    expect(asFragment()).toMatchSnapshot()
  })

  // it('with disabled', () => {
  //   const { asFragment } = render(<InputField disabled />)
  //   expect(asFragment()).toMatchSnapshot()
  // })
})
