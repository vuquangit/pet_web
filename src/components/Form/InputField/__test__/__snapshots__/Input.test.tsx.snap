// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Field component renders the component 1`] = `
<DocumentFragment>
  <div>
    <label
      class="block text-sm font-medium leading-6"
    >
      Input label
    </label>
    <div
      class="relative mt-2"
    >
      <input
        class="input-default"
        data-cy="input-cy"
        placeholder="Input placeholder"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;
